airbyte-cdk/python/airbyte_cdk/entrypoint.py

```diff
@@ -232,10 +236,25 @@ def _emit_queued_messages(self, source: Source) -> Iterable[AirbyteMessage]:
 def launch(source: Source, args: List[str]) -> None:
     source_entrypoint = AirbyteEntrypoint(source)
     parsed_args = source_entrypoint.parse_args(args)
-    for message in source_entrypoint.run(parsed_args):
-        # simply printing is creating issues for concurrent CDK as Python uses different two instructions to print: one for the message and
-        # the other for the break line. Adding `\n` to the message ensure that both are printed at the same time
-        print(f"{message}\n", end="", flush=True)
+    messages = iter(source_entrypoint.run(parsed_args))
+
+    if parsed_args.flush == 0:
+        for message in messages:
+            print(f"{message}\n", end="", flush=True)
+        return
+
+    while True:
+        with contextlib.closing(io.StringIO()) as buffer:
+            try:
+                with contextlib.redirect_stdout(buffer):
+                    end_time = time.monotonic() + parsed_args.flush
+                    while time.monotonic() < end_time:
+                        message = next(messages)
+                        buffer.write(f"{message}\n")
+            except StopIteration:
+                break
+            finally:
+                print(buffer.getvalue(), end="", flush=True)
```

Comment:
why do we need while cycle here with try...except?

can we use the same StringIO object and use `seek` and `truncate` method?

Reply:
Another concern I have is about concurrency.` io.StringIO()` is not thread-safe; maybe we should consider using `thread.Lock()` ?